another relational storage refactoring

shayhatsor opened the pull request:
This continues the work started with #1488. 
The main changed are:
* Minimized the SQL scripts (MySQL & SQL Server), fixed some minor bugs along the way.
* Refactored `RelationalOrleansQueries` to be short and readable.
I hope @veikkoeeva will have some time to review.

shayhatsor commented:
Some notes about the SQL scripts refactoring:
* It still follows the extended membership protocol, although it doesn't use a row version. I'm pretty sure that optimistic concurrency on the table version is enough. That's an optimization which is relevant to other membership implementations. @gabikliot what do you think?
* In the membership table I didn't split the suspecting silos to a new table, because usually there would be just 2 entries there. so `SuspectTimes` is a column that holds structured data - which in this case, seems better to me.

shayhatsor commented:
@sergeybykov, something is wrong with the build. please fix. thanks

shayhatsor commented:
a note about the changes in `SqlMembershipTable` and `SqlStatisticsPublisher`. I've added the `await` keyword in order for the logging to work as expected, previously it didn't catch the exceptions. The thing is that I'm not sure that there's a need for these added logs. I think that Orleans does a pretty good job in logging in general. An example of something that does zero logging by itself is `SqlReminderTable`. @sergeybykov, what is the preferred approach when implementing plugins ? I'd prefer we enhance Orleans` logs (if needed) and keep the plugins just about code.

jdom commented:
@dotnet-bot retest this please

sergeybykov commented on test/TesterInternal/MembershipTests/AzureMembershipTableTests.cs:
```diff
@@ -11,15 +13,13 @@ namespace UnitTests.MembershipTests
             TraceLogger.AddTraceLevelOverride("OrleansSiloInstanceManager", Severity.Verbose3);
             TraceLogger.AddTraceLevelOverride("Storage", Severity.Verbose3);
-
-            TestUtils.CheckForAzureStorage();
```
Why remove this?

shayhatsor replied:
Because these tests stopped working in the move to xunit. This needs to run before all other tests in the class. so I fixes it like [this](https://github.com/shayhatsor/orleans/blob/RefactorSql/test/TesterInternal/MembershipTests/AzureMembershipTableTests.cs#L16)

sergeybykov commented:
@shayhatsor Can you elaborate on the plugin question? I'm afraid I'm missing the context.

Looking at the PR, I see that you also refactored the membership table tests. Could you shed a light on the rational.

veikkoeeva commented:
@shayhatsor I'll take a look within the next 24 hours (it's now almost midnight here). With a very quick scan, I wonder, for instance in `InsertMembershipKey` what happens if the insert to `OrleansMembershipTable` fails, but update to `OrleansMembershipVersionTable` succeeds (due to a failed WHERE condition)? In that case `SET @ROWCOUNT = @@ROWCOUNT;` would be something else than `0` and it would be reported as successful operation. Then the other way around at `UpdateMembershipKey`.

shayhatsor commented:
@sergeybykov about the plugin question, what I meant is that I'd rather have this:
```csharp
public Task<MembershipTableData> ReadRow(SiloAddress key)
{
   return orleansQueries.MembershipReadRowAsync(deploymentId, key);
}
```
than this:
```csharp
public async Task<MembershipTableData> ReadRow(SiloAddress key)
{
    if (logger.IsVerbose3) logger.Verbose3(string.Format("SqlMembershipTable.ReadRow called with key: {0}.", key));
    try
    {
        return await orleansQueries.MembershipReadRowAsync(deploymentId, key);                
    }
    catch(Exception ex)
    {
        if (logger.IsVerbose) logger.Verbose("SqlMembershipTable.ReadRow failed: {0}", ex);
        throw;
    }
}
```
When I mentioned plugins I meant `IMembershipTable`, `IReminderTable` etc.
I think that any implementation should rely on Orleans to log the method invocations/exceptions.

shayhatsor commented:
@sergeybykov about :
> Looking at the PR, I see that you also refactored the membership table tests. Could you shed a light on the rational.

I haven't changed any logic, I've just converted the Asserts to xunit, and cleaned up a bit. I did add a new test method `GetGateways`. With it we have now complete test coverage of all relevant interfaces.
I should mention to all tests pass on Azure, SQL Server, MySQL and ZooKeeper.
Also, I think that we should add the "functional" trait to the Azure and Sql Server tests.

veikkoeeva commented:
@shayhatsor The SQL Server diff is too large to show in GH, so I just record here that when using the `@@ROWCOUNT` in `WHERE` in the SQL Server version, it might be clearer to write it as
```sql
SET @rowCount = @@ROWCOUNT;
IF @ROWCOUNT > 0
    UPDATE OrleansMembershipVersionTable
    SET
	Timestamp = GETUTCDATE(),
	Version = Version + 1
	WHERE
		DeploymentId = @deploymentId AND @deploymentId IS NOT NULL
		AND Version = @version AND @version IS NOT NULL
		AND @@ROWCOUNT > 0;

    SET @rowCount = @@ROWCOUNT;	
    IF @ROWCOUNT = 0
        ROLLBACK TRANSACTION
    ELSE
	COMMIT TRANSACTION
    
    SELECT @ROWCOUNT;
```
Nitpicking that I think it is more customary to write the variable names as in C#, but I suppose there is no single style widely used and especially not across all vendors and consistency is the key here and in this PR in particular.

veikkoeeva commented:
> about the plugin question, what I meant is that I'd rather have this:

@sergeybykov  This is related to the discussion long-time ago while refactoring the relational implementation and if we should unify the exception handling across all current and future membership table providers. That is, if the membership providers would just propagate the exceptions out and the Orleans core framework would log them. That would unify the key points of logging by construction. I think the concensus we reached back then was that we don't want to refactor it, at least not then, since it would have not OK for the ATS implementation to change. 

@shayhatsor A history explanation. :)

veikkoeeva commented on src/OrleansSQLUtils/Storage/DbExtensions.cs:
```diff
@@ -175,10 +177,10 @@ public static TValue GetValue<TValue>(this IDataRecord record, string fieldName)
         /// <exception cref="IndexOutOfRangeException"/>
         /// <remarks>This function throws if the given <see paramref="fieldName"/> does not exist.</remarks>        
-        public static async Task<TValue> GetValueAsync<TValue>(this DbDataReader record, string fieldName)
+        public static async Task<TValue> GetValueAsync<TValue>(this DbDataReader record, string fieldName, CancellationToken cancellationToken=default(CancellationToken))
```
There isn't a matching XML comment.

shayhatsor replied:
will fix

veikkoeeva commented on src/OrleansSQLUtils/Storage/DbStoredQueries.cs:
```diff
@@ -0,0 +1,406 @@
+        /// <summary>
+        /// A query template to retrieve gateway URIs.
+        /// </summary>        
+        public readonly string GatewaysQueryKey;
```
As these are public `readonly` fields, could they, should they, be readonly propery accessors (
`public string GatewaysQueryKey { get; }` would look more idiomatic.

As these are internal to Orleans implementation, could they be internal?

shayhatsor replied:
I've a huge fan of `readonly` fields. I'm not against property accessors, since in C# 6 they act the same as `readonly` fields. So if you're more inclined toward properties, I'm OK with it. about making these `internal`, well the class is `internal` so I'm not sure if that would be beneficial. But again, I'm OK with changing to internal.

veikkoeeva commented on src/OrleansSQLUtils/Storage/DbStoredQueries.cs:
```diff
@@ -0,0 +1,406 @@
+        }
+
+#pragma warning restore 0649
+        internal class Params
```
A matter of personal taste perhaps that I'd prefer these classes to become as first so I've read them before they're being used. It looks like the .NET and C# naming conventions are not used here (properties with small capital letter etc.).

shayhatsor replied:
I wondered about the naming. Maybe it'd be better to just name everything C# style, the parameters too. So we'd have the same parameter and column names, IMO will be easier to follow. The `DbStoredQueries` class is now the only one that "knows" about parameters, queries and column names. the nested `Params` class is responsible for all parameter definitions, and since a lot of them are the same in many queries, I've put them together and by that - added a bit of more indirection, but we gained separation of concerns.

veikkoeeva replied:
I'm OK with the idea of modifying the SQL side of things. Might be a slight incompatibility break, but not problematic and to me personally it's more important to converge more to the final design as quickly as possible. One's mileage may vary.

shayhatsor replied:
In you opinion - what's the delta between my proposed schema and the final design?

veikkoeeva replied:
I might not understand the question correctly, but I can clarify my point and maybe answers yours too. I'm not sure if your changes introduce backwards incompatible changes. If they do, they don't look like being big and I'm personally OK with them. Also it looks to me there isn't such a problem currently as there's no "business data" involved (i.e. storage provider, streams or reminders).

As for the future and "final design", I contemplate taking on a state provider this month as a new feature -- with sharding. This shouldn't touch anything currently. After state maybe statistics logging or streams. Of these, statistics logging would be backwards incompatible.  One potentially breaking change will be introduction of geo-distributed structures and if it is non-backwards compatible, well, I don't know if migration should be attempted.

shayhatsor replied:
The changes I'm making in this PR aren't backward compatible. I've actually optimized each table and query in order to simplify and remove redundancies. About the statistics tables, I haven't made any changes. I don't see a problem changing the statistics tables in the future even though it might be considered a "breaking change".

veikkoeeva commented on src/OrleansSQLUtils/Storage/DbExtensions.cs:
```diff
@@ -70,19 +71,20 @@ public static class DbExtensions
+        /// <param name="dbType">the <see cref="DbType"/> of the parameter.</param>
         /// <returns>A parameter created using the given arguments.</returns>
-        public static IDbDataParameter CreateParameter<T>(this IDbCommand command, ParameterDirection direction, string parameterName, T value, int? size = null)
+        public static void AddParameter<T>(this IDbCommand command, string parameterName, T value, ParameterDirection direction = ParameterDirection.Input, int? size = null, DbType? dbType=null)
```
Now it becomes more difficult to create parameters without adding them to the factory that creates them. I wonder then win here.

shayhatsor replied:
I wondered about this, I'll keep the `CreateParameter` method and add an `AddParameter` one.

veikkoeeva commented on src/OrleansSQLUtils/Storage/DbStoredQueries.cs:
```diff
@@ -0,0 +1,406 @@
+
+        internal static bool GetSingleBooleanValue(IDataRecord record)
+        {
+            if (record.FieldCount != 1) throw new ArgumentException("expected a single column", nameof(record));
```
I believe the exception message here should be capitalized. This is a bit tricky one to throw as this checks if the return value was what what excepted and throws if not. If this is desired, I think `InvalidOperationException` is the way to go. That's what, e.g., LINQ's `Single` returns when there isn't exactly one element in the sequence (i.e. kind of dynamic assert): https://msdn.microsoft.com/en-us/library/bb155325(v=vs.100).aspx.

shayhatsor replied:
will fix.

veikkoeeva commented on src/OrleansSQLUtils/CreateOrleansTables_MySql.sql:
```diff
@@ -36,509 +27,415 @@ Implementation notes:
+
+	SET _ROWCOUNT = ROW_COUNT();
+
+	IF _ROWCOUNT = 0
```
This goes also to SQL Server, but I believe it would be more clear to explicitly `CAST` the integer here before returning it or commenting it will be interpreted as a `bool` in the calling code. There'll be dragons here in that SQL Server uses `BIT`, which is a different bit than that in MySQL, and MySQL equivalent might be `BOOL`, which is an alias to `TINYINT(1)` (if I remember correctly). Some more info: http://troels.arvin.dk/db/rdbms/#data_types-boolean.

shayhatsor replied:
I documented [here](https://github.com/shayhatsor/orleans/blob/RefactorSql/src/OrleansSQLUtils/CreateOrleansTables_SqlServer.sql#L17) how Orleans interprets boolean values. i will add comments where this actually comes to play in the sql scripts. 
Keep in mind that my goal was to make the diff between the MySql and SQL server scripts as small as possible(do the diff with something like beyond compare). Things are a bit less explicit now, but IMO it makes it easier to introduce new providers (no need to bother converting to boolean since Orleans does that for you). this is also relevant to your [previous comment](https://github.com/dotnet/orleans/pull/1617?_pjax=%23js-repo-pjax-container#issuecomment-203094011)

veikkoeeva replied:
Ah, true. Somehow I didn't think about it though I saw it. Also I think I came across a bit differently than I intended to. What I tried to convey that maybe having a comment here this is interpreted as a `bool` by the framework is the best since the concept of `bool` is a mess in SQL. Maybe it is enough to have it in the initial comment, a judgement call.

veikkoeeva replied:
I'll record one oddity I just remembered, empty strings and nulls are treated somewhat interchangeably in Oracle: https://asktom.oracle.com/pls/asktom/f?p=100:11:0::::P11_QUESTION_ID:5984520277372. I have acquired the impression that's because of null terminated character arrays in C (and C++) and is due to historical reasons.

veikkoeeva commented:
It's one o'clock at night, so I doubt I get anything coherent out anymore. :) Nothing that really pokes one's eye (for what I see). I'll note that the relational tests used to be in the low milliseconds range but where run now in multiple seconds, on occasion cancelled due to a timeout. Then some of the statistics tests are still failing, but that wasn't a concern in this PR.

veikkoeeva commented:
Upon running the category `SqlServer` another time, plenty of tests fail. Here is one stack trace
```
Test Name:	UnitTests.StorageTests.SQLAdapter.RelationalStoreTests.CancellationToken_SqlServer_Test
Test FullName:	UnitTests.StorageTests.SQLAdapter.RelationalStoreTests.CancellationToken_SqlServer_Test
Test Source:	C:\projektit\Orleans\test\TesterInternal\StorageTests\RelationalStoreTests.cs : line 95
Test Outcome:	Failed
Test Duration:	0:00:00,013

Result StackTrace:	
kohteessa Xunit.Skip.If(Boolean condition, String reason)
   kohteessa UnitTests.StorageTests.SQLAdapter.RelationalStoreTests.CancellationTokenTest(RelationalStorageForTesting sut, TimeSpan timeoutLimit) tiedostossa C:\projektit\Orleans\test\TesterInternal\StorageTests\RelationalStoreTests.cs:rivillä 183
   kohteessa UnitTests.StorageTests.SQLAdapter.RelationalStoreTests.<CancellationToken_SqlServer_Test>d__12.MoveNext() tiedostossa C:\projektit\Orleans\test\TesterInternal\StorageTests\RelationalStoreTests.cs:rivillä 96
--- Pinon jäljityksen loppu aiemmasta sijainnista, jossa palautettiin poikkeus ---
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   kohteessa Xunit.Sdk.TestInvoker`1.<>c__DisplayClass46_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() tiedostossa C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:rivillä 227
--- Pinon jäljityksen loppu aiemmasta sijainnista, jossa palautettiin poikkeus ---
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   kohteessa Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext() tiedostossa C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:rivillä 48
--- Pinon jäljityksen loppu aiemmasta sijainnista, jossa palautettiin poikkeus ---
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   kohteessa Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext() tiedostossa C:\BuildAgent\work\cb37e9acf085d108\src\xunit.core\Sdk\ExceptionAggregator.cs:rivillä 90
Result Message:	Xunit.SkipException : Database was not initialized correctly
```

shayhatsor commented:
@veikkoeeva , I have changed the tests in #1488. They are a lot more thourough then before. I test everything several times and concurrently, so they might take a bit more time. Also, now, I'm creating a clean db for each test, it's a very small performance penalty, for the benefit of having all set-up and tear-down logic in one place, in the constructor.

veikkoeeva commented:
@shayhatsor Something else this time: https://gitter.im/dotnet/orleans?at=56fb05258f5147e119f19505. :) I better sleep now.

shayhatsor commented:
When i said "all tests pass" I meant all the tests that I've added or edited. As I mentioned [before](https://github.com/dotnet/orleans/pull/1488#issuecomment-187238185) the test configurations aren't aligned. I didn't put the effort into figuring out what's needed for those because they test things at a higher level (like silo configuration), and I'm confident that the improved tests from #1488 cover everything I touched. 
These tests should pass:
CancellationToken_MySql_Test
CancellationToken_SqlServer_Test
MembershipTable_Azure_GetGateways
MembershipTable_Azure_Init
MembershipTable_Azure_InsertRow
MembershipTable_Azure_ReadAll_EmptyTable
MembershipTable_Azure_ReadAll_Insert_ReadAll
MembershipTable_Azure_ReadRow_Insert_Read
MembershipTable_Azure_UpdateRow
MembershipTable_Azure_UpdateRowInParallel
MembershipTable_MySql_GetGateways
MembershipTable_MySql_Init
MembershipTable_MySql_InsertRow
MembershipTable_MySql_ReadAll_EmptyTable
MembershipTable_MySql_ReadAll_Insert_ReadAll
MembershipTable_MySql_ReadRow_Insert_Read
MembershipTable_MySql_UpdateRow
MembershipTable_MySql_UpdateRowInParallel
MembershipTable_SqlServer_GetGateways
MembershipTable_SqlServer_Init
MembershipTable_SqlServer_InsertRow
MembershipTable_SqlServer_ReadAll_EmptyTable
MembershipTable_SqlServer_ReadAll_Insert_ReadAll
MembershipTable_SqlServer_ReadRow_Insert_Read
MembershipTable_SqlServer_UpdateRow
MembershipTable_SqlServer_UpdateRowInParallel
RemindersTable_Azure_Init
RemindersTable_Azure_ReminderSimple
RemindersTable_Azure_RemindersParallelUpsert
RemindersTable_Azure_RemindersRange
RemindersTable_MySql_Init
RemindersTable_MySql_ReminderSimple
RemindersTable_MySql_RemindersParallelUpsert
RemindersTable_MySql_RemindersRange
RemindersTable_SqlServer_Init
RemindersTable_SqlServer_ReminderSimple
RemindersTable_SqlServer_RemindersParallelUpsert
RemindersTable_SqlServer_RemindersRange
SqlStatisticsPublisher_MySql_Init
SqlStatisticsPublisher_MySql_ReportMetrics_Client
SqlStatisticsPublisher_MySql_ReportMetrics_Silo
SqlStatisticsPublisher_MySql_ReportStats
SqlStatisticsPublisher_SqlServer_Init
SqlStatisticsPublisher_SqlServer_ReportMetrics_Client
SqlStatisticsPublisher_SqlServer_ReportMetrics_Silo
SqlStatisticsPublisher_SqlServer_ReportStats
Streaming_MySql_Test
Streaming_SqlServer_Test

shayhatsor commented:
about the stack trace you've posted, it seems that the db isn't up or something. AFAIK, i haven't touched anything that may bother with the test you've mentioned (`CancellationToken_SqlServer_Test`). It does pass on my end.

shayhatsor commented:
@veikkoeeva, about [your comment about error logging](https://github.com/dotnet/orleans/pull/1617?_pjax=%23js-repo-pjax-container#issuecomment-203097510). This means I should add logging to `SqlReminderTable`, because currently it has none. I think it'd be better to create a wrapper class for `IMembershipTable`, `IReminderTable` etc. that will do the logging. It would be beneficial to all providers (will not used on Azure since it has its own logging inside).

veikkoeeva commented:
@shayhatsor If you can run `SqlServer` category, could you check if it is dog slow for you too? The MySQL tests for me are quite fast, not problem there. Also, I wonder if `isSynchronousAdoNetImplementation` (or `IsSynchronousAdoNetImplementation` as I'm looking into master now) is reported correctly. It might be that I've just blundered something now, but it looks to me it's the wrong way around. I don't remember introducing this a serious performance degradation while doing the other work, but that might be possible too... But not to to sleep, in earnest.

shayhatsor commented:
@veikkoeeva, yep i see your point, they are slower. I've profiled the tests and it seems that sql server doesn't like to have the db created and destoyed on every method, so the performance penalty is bigger than i said [here](https://github.com/dotnet/orleans/pull/1617#issuecomment-203138229) .will fix.

veikkoeeva commented:
@shayhatsor OK, thanks a bunch. I don't think there is a need to create and re-create the db every time, just crop the tables (or delete the contents therein). The issue might be that as I made the initial size bigger the last time around (to 20 MiB) due to the streaming tests, it creates more churn. Just in case: if we'd need to seed the DB with a lot of data, a very quick method would be to make an initial backup (with a temp name to a temp folder) and restore it when needed (a diff backup takes time some hundreds milliseconds even on hundreds of megabytes of a db).

I also got the tests to run faster, in some seconds range, by switching `.Wait()` calls to `.GetAwaiter().GetResult()` and putting `<add key="xunit.shadowCopy" value="false"/>` to `App.config` files. Need to get kids to kindergarden now, sorry for semi-hijacking this thread and thanks.

veikkoeeva commented:
@shayhatsor Maybe we can use [[ICollectionFixture]](https://gist.github.com/bradwilson/8423477) to denote which group of tests need to be run serially. I think at most this would mean arranging some random deployment for different collections (e.g. testing membership protocl in parallel with reminders in different deployments, though in the same DB).

Maybe relevant also there is already [[SkippableFact]](https://github.com/xunit/samples.xunit/blob/master/DynamicSkipExample/Samples.cs)  used. With this it's possible to check dynamically if some condition holds and mark tests skipped accordingly. For instance, I don't have `ZooKeeper` installed locally, so it would be nice if those tests would be skipped automatically when I "run all". And consequently not skipping when an environment is installed. I realize we still need categories and writing a "detector" is a more involved business than this PR (there is one for the Azure emulator), but it looks to me it makes to sense to consider this when arranging the tests.

I'm at work currently and you know better, but checking quickly on another machine the tests are rather slow and indeed it looks like as if the DB will be destroyed and re-built (not just cropping the tables) on every test. LocalDb isn't that a fast engine and it shows.

veikkoeeva commented:
@shayhatsor An additional note about that [thanks a bunch](https://en.wiktionary.org/wiki/thanks_a_bunch), I meant it in the less common, unsarcastic meaning.

shayhatsor commented:
about [this comment](https://github.com/dotnet/orleans/pull/1617#issuecomment-203609477), It didn't even cross my mind. It's all good.

sergeybykov commented:
@shayhatsor @veikkoeeva Do you guys think this will be ready to get included in 1.2.0? It'll probably take us another week before we are ready for a beta.

shayhatsor commented:
@sergeybykov, it might be ready by then. I'll update in a few days.

veikkoeeva commented:
@sergeybykov I don't see anything fatal here, more like synchronizing minds. Less code is good.

Related to tests refactoring, I would be thankful if the DB tests would be again in the milliseconds range. The working strategy here is not to recreate the whole database on every test. Even just cropping the tables is much, much faster, but even that may not be necessary. I believe those two linked attributes and/or using random deployments and running concurrently should work. Maybe a point of coordination is on how should storage tests being run in general.

shayhatsor commented:
@veikkoeeva, I've addressed all the issues mentioned. Please review, thanks

veikkoeeva commented:
I will, @shayhatsor. Tomorrow, so I would estimate I should be ready in 26 hours.

veikkoeeva commented on src/OrleansSQLUtils/Storage/DbStoredQueries.cs:
```diff
@@ -0,0 +1,425 @@
+        }
+
+        /// <summary>
+        /// the query that's used to get all the stored queries.
```
A nit: these start with a small letter.

shayhatsor replied:
which ones should be capitalized ? I'm note sure I understand

shayhatsor replied:
oh I get it, I thought you're talking about the code, will fix

veikkoeeva replied:
Ah, "the query that's used to get all the stored queries.". Most (if not all, I'm not sure) comments are capitalized and end with a full stop. I don't know if there's a guidance for this in Orleans. So as to be clear, I go mainly how others go, but I prefer full sentences using proper writing (I explain it to myself it shouldn't at least hurt for trying to be concise and use proper language).

veikkoeeva commented on src/OrleansSQLUtils/Storage/RelationalStorage.cs:
```diff
@@ -286,7 +264,7 @@ private bool IsSynchronousAdoNetImplementation
 
                     Task<Tuple<IEnumerable<TResult>, int>> ret;
-                    if(IsSynchronousAdoNetImplementation)
+                    if(isSynchronousAdoNetImplementation)
```
@shayhatsor I left an unfortunate bug here the last time. Ultimately this capability is detected with this piece of code:

```
public static bool IsSynchronousAdoNetImplementation(string adoNetProvider)
        {
            //Currently the assumption is all but MySQL support DbCommand cancellation.            
            return !adoNetProvider.Equals(AdoNetInvariants.InvariantNameMySql, StringComparison.OrdinalIgnoreCase);
        }
```
As you can see, it shouldn't be negated.

shayhatsor replied:
I'm fixing it now

veikkoeeva replied:
:+1:

veikkoeeva commented on src/OrleansSQLUtils/Storage/RelationalOrleansQueries.cs:
```diff
@@ -382,886 +255,185 @@ public static Task<RelationalOrleansQueries> CreateInstance(string invariantName
 
-
         /// <summary>
         /// deletes all reminders rows of a service id
```
Nit: starts with a small letter, should probably be capitalized and perhaps a point added.

shayhatsor replied:
will fix

veikkoeeva commented on test/TesterInternal/MembershipTests/MembershipTableTestsBase.cs:
```diff
@@ -1,72 +1,76 @@
 
             logger.Info("DeploymentId={0}", deploymentId);
 
+            lock (fixture)
```
@shayhatsor You may know this, but locking widely available object bode ill luck. One explanation: http://stackoverflow.com/questions/11775205/why-is-it-a-bad-practice-to-lock-the-object-we-are-going-to-change. This looks fishy in that there are virtual calls in constructor and in addition the given parameter is used in a lock. Might not hurt now -- or ever -- but raises a spectre of suspicion upon reading. I would consider changing this, but on the other hand, I gather the test code is somewhat in flight currently and it might be beneficial to get this in. It might also be the Orleans maintainers are OK with this. @jdom, I think, is moving the test refactoring forwards. @sergeybykov may have an opinion.

shayhatsor replied:
well, in this case it shouldn't make a difference, but to make it readable to anyone but me, i'll fix it.

veikkoeeva commented:
@shayhatsor Apart for the minor things, looks good to me. I'll go to sleep now and take a look at in the morning if there's something I can help before merging. Laters.

shayhatsor commented:
@sergeybykov, I believe this PR is ready for merge.

sergeybykov commented:
@shayhatsor @veikkoeeva Thanks a lot!
